### DISCUSSION TITLE
Github Action trigger on release not working if releases was created by automation?

### DISCUSSION THREAD

**Discussion by Geertvdc** [community/community#25281] [Actions]
<p>I’ve been working on some workflows that automatically build &amp; release my application using Github Actions.</p>
<p>i have 2 workflows:</p>
<p>1 <strong>build</strong> workflow -&gt; this workflow builds my solution and creates some artifacts (zip files). using the Hub cli i create a release and attach these zip files to the release.</p>
<p>source here: <a href="https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/build.yml" rel="nofollow noopener">https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/build.yml</a></p>
<p>1 <strong>release</strong> workflow -&gt; this workflow <em>should</em> trigger on a release that is created, download the zip files from the release and deploy them to my environments. however the release trigger is not triggered when the release is created.</p>
<p>source here: <a href="https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/release-prod.yml" rel="nofollow noopener">https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/release-prod.yml</a></p>
<p>i’ve tried different things to trigger the release workflow with just on: release to defining the types but results are the same. It just doesn’t trigger the workflow.</p>
<pre><code>on:
  release:
    types: [created, published]
</code></pre>
<p>When i create a new release by hand instead of using the hub cli it does trigger my workflow. Any ideas how come?</p>
<p>edit: i checked running <code>hub release create</code> by hand and then the release workflow just starts. there is something that if it is triggered inside another workflow that it doesn’t trigger it.</p>

**Comment by Geertvdc** [comment, marked as answer]
<p>I found the solution with help from <a class="mention" href="/ethomson">@ethomson</a> on <a href="https://twitter.com/GeertvdC/status/1183837353493893122" rel="nofollow noopener">twitter</a>. The thing I changed is instead of passing in the default GITHUB_TOKEN you have in your secrets during a workflow I’m using my personal PAT token.</p>

**Comment by elucidsoft** [comment]
<p>Thank you for this, I feel like this should be a bug…</p>

**Comment by nkt** [comment]
<p><span class="mention">@github</span> is this expected behavior?</p>

**Comment by Geertvdc** [comment]
<p><a class="mention" href="/nkt">@nkt</a> as you see in my answer <a class="mention" href="/ethomson">@ethomson</a> (who works for Github) mentioned that this is how they designed it (see tweet link above in my answer)</p>

**Reply by asgerjensen** [reply]
Great. Now just maintain that  pr-repo PAT on 400 repos with a max lifetime of one year...

**Comment by czerasz** [comment]
<p>Suboptimal from the users perspective :slightly_frowning_face:</p>

**Comment by ghost** [comment]
<p>Coming from the hell of Jenkins and Gitlab CI, I’ve discovered the new depths of hell I didn’t even think were possible.</p>
<p>This “design” decision is a pinnacle of stupidity.</p>

**Comment by johnnyoshika** [comment]
<p>Well this is a surprise. For anyone else who’s wondering, I granted the <code>repo</code> permission when generating the personal access token in order to create a release within a GitHub Action:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/b/e/bea4012188e31f2d48d09707795dfb22b3088692.png" data-download-href="/uploads/short-url/rcu61KGG5MmcWzyakFA7hdLpnc6.png?dl=1" title="Screenshot 2022-02-24 212908" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/504505/181090062-b6edb3a4-829a-434a-8fcf-68681d84eeee.png" alt="Screenshot 2022-02-24 212908" data-base62-sha1="rcu61KGG5MmcWzyakFA7hdLpnc6" width="690" height="226" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/b/e/bea4012188e31f2d48d09707795dfb22b3088692_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">Screenshot 2022-02-24 212908</span><span class="informations">791×260 11.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by stefreak** [comment]
GitHub prevents workflows from running on events that were caused by other workflows to prevent unlimited recursion (See the docs: [Triggering a workflow from a workflow](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow))

If you create a release within a workflow, one alternative to run another workflow after the release has been created is to use the [`workflow_run`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_run) trigger. It has a maximum "recursion depth" to prevent unlimited recursion.

Using `workflow_run` is easier to maintain in the long run than to use personal access tokens to work around this issue.

*Note to the GitHub devs: I see that this behaviour makes sense, for example it makes it easier to develop workflows that edit issues and also run when issues are changed. Meanwhile it would be still great if GitHub devs could do something about the user experience here, as I lost a couple of hours today to learn about this behaviour; In the case of automatically created releases this was definitely contra-intuitive for me (and reading the tone in this discussion I am not alone).*

**Reply by stefreak** [reply]
I actually ended up calling a reusable workflow to deploy the software within the workflow that creates the release, instead of using the release trigger. That way I preserved all the information of the original push event in my github context

**Reply by warrenackerman** [reply]
The workflow_run approach does not allow enough filtering capabilities. It only allows a single condition of sorts. Calling from a job allows for more complex conditions to be met.

**Reply by fracampit** [reply]
The workflow_run approach also doens't help when the release process invloves several steps, that span outside the workflow.

for example, with [release please](https://github.com/googleapis/release-please), the process looks like this:
* create new branch to do some work
* create PR to the default branch
* merge PR
* release please runs on commits to the default branch
* release please sees the new commits from the PR merge and, if necessary (depending on the commits), creates a [release PR](https://github.com/googleapis/release-please?tab=readme-ov-file#whats-a-release-pr)
* merge the release PR
* release please then sees a new "special" commit on the default branch due to merging the release PR (more info on the special commit [here](https://github.com/googleapis/release-please/blob/main/docs/customizing.md#release-lifecycle-labels)), and creates a release

In this case, you can't just add a setp into the release-please workflow to make it deploy, because it would essentially deploy on every single commit to the default branch, which is obviously wrong given the fact that releases are only created with the flow described above.

**Reply by ykamothi-sh** [reply]
Took me hours to finally get to this, this is the answer!

**Comment by ghost** [comment]
Yeah, this was exactly my experience with Microsoft support: stupid, uncooperative, will rigmarole you with unhelpful replies until you just give up.

And when the problem is obvious, they will come up with all sorts of imaginary reasons why not fix it. And the reason for this is: support can do nothing about how the product is made, they can only color between the lines of what they were told to do, the R&D don't care, and in most cases have no idea how the product is used or how it could be potentially used.

It's funny that this Github Action thing was supposed to be the "killer feature" that would vendor-lock Github users into using more proprietary Microsoft stuff, but as with all other Microsoft "solutions", it's just so pathetically bad, and reeks of Microsoft from many miles away... It's ironic that no matter how they try, they always get Windows Vista experience.

**Reply by stefreak** [reply]
If you are referring to my post here, I want to make clear that I am not associated with Microsoft or GitHub at all in any way. I also had the same problem you faced.

Every system has its limitations, some more, some less, but I am absolutely sure that the people at Microsoft are working hard to improve things. It is really hard to change big systems used by millions of people and it can sometimes take a long time. Nobody is smart enough to come up with just the right solutions for every problem from the beginning.

**Comment by johnnyoshika** [comment]
@wvxvw That's rather harsh. It would be much more productive to give constructive feedback instead of insulting Microsoft. While I don't agree with everything Microsoft does and I was surprised to see that I had to create a personal access token to trigger the on.release.created event, I also appreciate the contributions they make to the open source community and free developer tools. I also understand that I have a choice not use their services if I don't want to and you do as well. You may want to consider exercising that option.

**Comment by ringerc** [comment]
So ... "on release" doesn't mean "on release", it means "on manually created release".

That's ... er ... *special*. Because it's already so easy to compose github workflows into sensible, modular units.

See also https://github.com/community/community/discussions/16244

**Comment by guhcampos** [comment]
This is a very unfortunate behavior, as `workflow_run` does not allow some common use cases. My example:

I'm using the [helm/chart-releaser-action](https://github.com/helm/chart-releaser-action) to trigger a github release when a chart is modified. It will only trigger a new release if `Chart.yaml` has a new `version`. I then expected to run a new post-release workflow with:

```
on:
  release:
    types: [published]
```

... which does not work because of this design choice.

Using `workflow_run` would not work, as the release workflow **always runs** but **does not always trigger a release** - my post-release workflow should only run when the release workflow has run **and created a release**. 

I will have to use one of the multiple very dirty workarounds in this case. It does not make it impossible, but rather confusing.

**Reply by FloSchick** [reply]
@guhcampos i have a similar issue with the terraform-module-releaser. I want renovate to run right after a new release was published by the other action. Have you found any solution for this?

**Reply by guhcampos** [reply]
Hmm unfortunatelly this was way too long ago that I don't remember and I switched companies since, so can't really tell.

**Comment by damianesteban** [comment]
Thanks for this, it took me forever to figure out.

**Comment by fracampit** [comment]
One aspect of this design choice that I find puzzling is the focus on preventing infinite loops. While the limitation on `GITHUB_TOKEN` workflows aims to address this, users can still inadvertently create infinite loops using workflows triggered by personal access tokens (`PAT`s). Since it’s impossible to completely prevent infinite loops in a system where users can write custom workflows, this restriction feels unnecessarily strong and counterproductive.

The limitation creates a significant obstacle for automating releases and deployments, forcing teams to adopt workarounds like using a team member’s personal token, which raises visibility and auditing concerns.

I believe this decision could be revisited, considering:

* The infinite loop issue can’t be fully mitigated in a system with user-defined workflows.
* The restriction adds unnecessary friction to legitimate use cases like automating releases.
* An opt-in mechanism or a flag to allow trusted workflows triggered by `GITHUB_TOKEN` could be a more balanced approach, giving teams flexibility while keeping safeguards in place for those who need them.


As a potential workaround for the current limitation, I’m considering using a scheduled workflow. This workflow would periodically check for new releases via the GitHub API and trigger deployments when a new release is detected. While this approach sidesteps the limitation, it comes with trade-offs:

* It introduces a delay between the release creation and deployment.
* It’s less efficient as it requires periodic checks, consuming additional workflow minutes.
* State tracking is required to avoid redeploying the same release multiple times.

While not ideal, this workaround might be a temporary solution until a more robust mechanism is provided. I’d love to hear feedback on whether others have tried this approach or have alternative ideas.

**Reply by cabbiepete** [reply]
fwiw I've worked around the issue by calling my deploy workflow after creating the github release.

I've left the trigger on release creation for if a release is created manually.

So far all works ok.

**Comment by Chiassonastraypatrick** [comment]
> <p>I’ve been working on some workflows that automatically build &amp; release my application using Github Actions.</p>
> <p>i have 2 workflows:</p>
> <p>1 <strong>build</strong> workflow -&gt; this workflow builds my solution and creates some artifacts (zip files). using the Hub cli i create a release and attach these zip files to the release.</p>
> <p>source here: <a href="https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/build.yml" rel="nofollow noopener">https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/build.yml</a></p>
> <p>1 <strong>release</strong> workflow -&gt; this workflow <em>should</em> trigger on a release that is created, download the zip files from the release and deploy them to my environments. however the release trigger is not triggered when the release is created.</p>
> <p>source here: <a href="https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/release-prod.yml" rel="nofollow noopener">https://github.com/XpiritBV/beer-xchange/blob/master/.github/workflows/release-prod.yml</a></p>
> <p>i’ve tried different things to trigger the release workflow with just on: release to defining the types but results are the same. It just doesn’t trigger the workflow.</p>
> <pre><code>on:
>   release:
>     types: [created, published]
> </code></pre>
> <p>When i create a new release by hand instead of using the hub cli it does trigger my workflow. Any ideas how come?</p>
> <p>edit: i checked running <code>hub release create</code> by hand and then the release workflow just starts. there is something that if it is triggered inside another workflow that it doesn’t trigger it.</p>

Well release what you won't but at my last cort date I had a global resolution so no one can charge me for anything prior to that date so stop the fucking bullshit I'm un killable

**Comment by Googlom** [comment]
that's why you shouldn't sell your company to Microsoft
